test/common/webdriver_bidi.py

```diff
@@ -0,0 +1,453 @@
+# This file is part of Cockpit.
+#
+# Copyright (C) 2024 Red Hat, Inc.
+#
+# Cockpit is free software; you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation; either version 2.1 of the License, or
+# (at your option) any later version.
+#
+# Cockpit is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+
+"""BiDi session/API driver
+
+This directly talks the https://w3c.github.io/webdriver-bidi/ protocol from async Python,
+without any node/JS in between. The only dependencies are aiohttp, firefox, and/or
+chromedriver+chromium (or headless_shell).
+"""
+
+import asyncio
+import contextlib
+import json
+import logging
+import os
+import socket
+import sys
+import tempfile
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any
+
+import aiohttp
+
+log_proto = logging.getLogger("bidi.proto")
+log_command = logging.getLogger("bidi.command")
+
+
+JsonObject = dict[str, Any]
+
+
+class WebdriverError(RuntimeError):
+    pass
+
+
+@dataclass
+class LogMessage:
+    level: str  # like "info"
+    type: str  # usually "console"
+    timestamp: int
+    args: list[object]
+    text: str
+
+    def __init__(self, message_params: JsonObject) -> None:
+        self.level = message_params["level"]
+        self.type = message_params["type"]
+        self.timestamp = message_params["timestamp"]
+        self.args = message_params.get("args", [])
+        self.text = message_params["text"]
+
+    def __str__(self) -> str:
+        return f"> {self.level}: {self.text}"
+
+
+@dataclass
+class BidiSession:
+    ws_url: str
+    session_url: str
+    process: asyncio.subprocess.Process
+
+
+# Return port numbers that were free at the time of checking
+# They might be in use again by the time the function returns...
+def pick_ports(count: int) -> list[int]:
+    sockets: list[socket.socket] = []
+    ports: list[int] = []
+
+    for _ in range(count):
+        sock = socket.socket()
+        sock.bind(('127.0.0.1', 0))
```

Comment:
I *guess* that could fail.  And I *guess* that we would theoretically want to close all of the other sockets in that case.  But on the other hand, this is being pretty extremely pedantic for a case that's probably never going to actually happen.

Reply:
I wrapped that in a `finally:`.